ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1135,6 +1306,189 @@ private void applyDelete(MessageInfo messageInfo, RemoteReplicaInfo remoteReplic
+          + " and replicasToReplicatePerNode size " + remoteReplicaInfos.size() + " should be the same");
+    }
+
+    for (int i = 0; i < remoteReplicaInfos.size(); i++) {
+      RemoteReplicaInfo remoteReplicaInfo = remoteReplicaInfos.get(i);
+      ReplicaId localReplica = remoteReplicaInfo.getLocalReplicaId();
+      ReplicaId remoteReplica = remoteReplicaInfo.getReplicaId();
+      // Check if local replica and remote replica are leaders for their partition.
+      if (leaderBasedReplicationAdmin.isLeaderPair(localReplica, remoteReplica)) {
+        leaderReplicaInfosOutput.add(remoteReplicaInfo);
+        exchangeMetadataResponseListForLeaderReplicaInfosOutput.add(exchangeMetadataResponseList.get(i));
+      }
+    }
+  }
+
+  /**
+   * Returns list of remote replica infos whose missing blobs in their metadata response haven't arrived within
+   * time = replicationConfig.replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds.
+   * @param remoteReplicaInfos list of remote replica infos
+   * @return list of remote replica infos which have timed out due to no progress
+   */
+  List<RemoteReplicaInfo> getRemoteStandbyReplicasTimedOutOnNoProgress(List<RemoteReplicaInfo> remoteReplicaInfos) {
+
+    // Use case: In leader-based replication, standby replicas don't send replication GET requests for missing keys
+    // found in their metadata exchange and expect them to come from leader in local data center via intra-dc replication.
+    // However, if for any reason, their missing blobs never arrive via local leader, this is a safety feature to fetch
+    // the blobs themselves in order to avoid being stuck.
+
+    // Example scenario: For DELETE after PUT use case in remote data center, it is possible that standby replicas get
+    // only PUT record in its replication cycle (DELETE record will come in next cycle) while leader gets both
+    // PUT and DELETE together in its replication cycle. Due to that, leader doesn't fetch
+    // the deleted blob from remote data center and the blob is never replicated from leader to standby.
+    // As a result, the PUT record in standby's missing blobs set is never emptied.
+
+    // Time out period is configurable via replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds. If
+    // replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds == -1, this safety feature is disabled.
+
+    List<RemoteReplicaInfo> remoteReplicasTimedOut = new ArrayList<>();
+    if (replicationConfig.replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds != -1) {
+      for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
+        ExchangeMetadataResponse exchangeMetadataResponse = remoteReplicaInfo.getExchangeMetadataResponse();
+        if (exchangeMetadataResponse.hasMissingStoreMessages()
+            && (time.seconds() - exchangeMetadataResponse.metadataReceivedTimeSec)
+            > replicationConfig.replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds) {
```

Comment:
This condition can be improved. Instead of using metadata received timestamp. We can use the last timestamp when at least one messageInfo is removed from missing set. This is more reasonable for the method name "no progress", it may make a slow progress and timed out in middle way.